### PR TITLE
clients/v1: define user metadata RPCs

### PR DESCRIPTION
**Pull request by bobheadxi**
Part of https://linear.app/sourcegraph/issue/CORE-545/confirm-absence-of-api-for-writing-sams-user-metadata - see issue description

## Test plan

CI

### CODE REVIEW

File: clients/v1/clients.proto [hunk 1]
```diff
@@ -39,6 +40,21 @@ service UsersService {
...
+    // Cannot use 'sams_required_scopes' due to metadata-scope-specific scopes.
+    // See docstring for required scopes.
+  }
+  // UpdateUserMetadata updates the metadata of a SAMS user.
+  //
+  // Required scopes: 'sams::user.metadata::write' or metadata-scope-specific
+  // variant scope, such as 'sams::user.metadata.dotcom::write'
+  rpc UpdateUserMetadata(UpdateUserMetadataRequest) returns (UpdateUserMetadataResponse) {
+    // Cannot use 'sams_required_scopes' due to metadata-scope-specific scopes.
+  }
```
Comment (bobheadxi) [inline clients/v1/clients.proto:57]: @unknwon per AIP this would always `overwrite = true`, WDYT?

File: clients/v1/clients.proto [hunk 2]
```diff
@@ -102,6 +118,54 @@ message Role {
...
+  // The user's external ID.
+  //
+  // Behaviour: IDENTIFIER
+  string user_id = 1;
+  // The service slug.
+  //
+  // Behaviour: IDENTIFIER
+  string service = 2;
+  // Metadata associated with this user for the specified service.
+  map<string, string> metadata = 3;
```
Comment (unknwon) [inline clients/v1/clients.proto:143]: In SAMS, the value type is not strictly `string`, I think it's fine to be stricter here if it's too much trouble to support `any`.

CleanShot 2024-12-09 at 18 00 29@2x

**Review by unknwon — APPROVED** [review]
Approve to unblock!

File: clients/v1/clients.proto (hunk 1 above)
Comment (bobheadxi) [reply clients/v1/clients.proto:57]: yeah, this parameter:

image

https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph-accounts/-/blob/backend/internal/database/user_metadata.go?L47

File: clients/v1/clients.proto (hunk 2 above)
Comment (bobheadxi) [reply clients/v1/clients.proto:143]: ah, yes we can support a more arbitrary shape with https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/struct.proto, we've used this for telemetry

File: auth/clientcredentials/connectrpc.go [hunk 3]
```diff
@@ -170,9 +170,6 @@ func extractSchemaRequiredScopes(spec connect.Spec, extension *protoimpl.Extensi
 		return nil, errors.Newf("extension field %s not valid", extension.TypeDescriptor().FullName())
 	}
 	list := value.List()
-	if list.Len() == 0 {
-		return nil, errors.Newf("extension field %s cannot be empty", extension.TypeDescriptor().FullName())
-	}
```
Comment (bobheadxi) [inline auth/clientcredentials/connectrpc.go:175]: @unknwon FYI I had to add this to allow us to have RPCs that don't automatically check for a specific scope in the middleware; see the updated tests

File: scopes/metadata.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+)
+
+// PermissionToMetadataScope extracts the metadata scope from a permission.
+// Only valid for 'user.metadata'-prefixed permissions.
+//
+// Examples:
+//   - user.metadata => *
+//   - user.metadata.cody => cody
+//   - user.metadata.dotcom => dotcom
+func PermissionToMetadataScope(permission Permission) (string, error) {
```
Comment (bobheadxi) [inline scopes/metadata.go:15]: @unknwon FYI I moved this from SAMS internals

File: clients/v1/clients.proto (hunk 1 above)
Comment (unknwon) [reply clients/v1/clients.proto]: Yep, LGTM!

**Review by unknwon — APPROVED** [review]
🚀

File: clients/v1/clients.proto [hunk 5]
```diff
@@ -102,6 +119,49 @@ message Role {
...
 
+message GetUserMetadataRequest {
+  // The user's external ID.
+  string id = 1;
+  // Services is the list of service slugs to filter the metadata. "*" can be used
+  // to return metadata for all services.
+  //
+  // Requires the scope for the requested service(s); for example, '*' requires
+  // 'sams::user.metadata::read', and 'dotcom' requires 'sams::user.metadata.dotcom::read'.
+  repeated string services = 2;
```
Comment (bobheadxi) [inline clients/v1/clients.proto:130]: @unknwon I realize that we call this "service" internally but it does not (and probably should not) align with "scope services", i.e. `services.Service`. Should we change the naming here to be more generic, like "namespace"?

Comment (bobheadxi) [reply clients/v1/clients.proto]: Renamed in 69d985475b17b43465f24001c3af5899f662efd3

**Pull request merged**